### ISSUE TITLE
Hash confusion when multiple protocols are involved

### ISSUE THREAD

**Issue by vain** [jointwt/twtxt#408]
(Filing as a bug because it can break threading.)

There are some twtxt feeds that are available over multiple protocols. Mine, for example:

* https://www.uninformativ.de/twtxt.txt
* gopher://uninformativ.de/0/twtxt.txt

Or that of melyanna:

* https://tilde.club/~melyanna/twtxt.txt
* http://tilde.club/~melyanna/twtxt.txt

I suspect there are more.

I follow melyanna’s feed via https, while prologic’s pod at twtxt.net follows via http. For a certain twt, I got the hash `kjsxeka`, while twtxt.net got `zafd26q`: https://twtxt.net/conv/zafd26q And thus, threading broke: My client shows the entire thread, that you can see at https://twtxt.net/conv/zafd26q, as orphaned/dangling, because it does not know anything about `zafd26q`.

One idea that already came up was omitting the URL scheme when hashing. That *might* work, not sure, but it must be done carefully: Stripping the scheme from Gopher URLs would result in `uninformativ.de/0/twtxt.txt`, which is still very different from `www.uninformativ.de/twtxt.txt`. (My Gopher setup is probably too exotic, because not even the host name matches. But the general problem remains, you have to strip the Gopher item type, which is technically part of the URLs `path`.)

Another idea might be: Demand that feed URLs are universally unique, don’t serve the same feed over two protocols, end of discussion. 🤔 Not sure about that.

**Comment by der-lyse** [comment]
Another spontaneous idea: If the feed specifies `url` comments, the first of them can be used to create the twt hash. Maybe even introduce a dedicated setting for that, but I reckon that this would be even more fragile.

**Comment by JonLundy** [comment]
If the feed declares its authoritative URL we should use that. At that point all needed information is self contained to the feed file.

**Comment by JonLundy** [comment]
Another case I have seen is feeds with fragments at the end. Twtxt.net adds the nick as a fragment. And I have seen Hash's break because someone follows the one with a fragment. 

```
https://txt.sour.is/users/xuu/twtxt.txt
https://txt.sour.is/users/xuu/twtxt.txt#xuu
```

**Comment by der-lyse** [comment]
> Another case I have seen is feeds with fragments at the end. Twtxt.net adds the nick as a fragment.

Aha! I've seen mentions with anchors and was wondering why on earth the nicks were appended. I think this is wrong, since nicks are already included in the first part of a mention.

**Comment by prologic** [comment]
It was done as a workaround for the mobile app￼

**Comment by vain** [comment]
> If the feed declares its authoritative URL we should use that.

I could then claim to be your feed. I wonder what the implications of that are.

In the context of just twt hashes, it’s probably harmless. The worst I could do is intentionally break threading (however, a typo or otherwise malformed URL in my feed metadata would also break threading, unintentionally).

In a broader context, it’s probably not a good idea to blindly trust URLs mentioned in feed metadata. :) If we start doing that, the URL might get used for something else later on and suddenly we have an impersonation issue.

Which leads me to the question: Should we treat `url =` lines as *redirects*? That way, I can’t pose as someone else, I can just redirect people there (I can do that via HTTP redirects as well).

**Comment by JonLundy** [comment]
What if it only processes if the meta url matches the fetched URL? So if it is different it has to fetch that one? I guess is that what you are saying?

**Comment by JonLundy** [comment]
really it would be better if there were signatures some how so it can just be signed and we know for sure its not impersonated

**Comment by prologic** [comment]
But how do we develop and establish trust for signatures in the first place? That's quite hard to do in a distributed system I think￼

**Comment by prologic** [comment]
What if we just make sure that the URL matches and drop the scheme as part of the hush?￼

**Comment by vain** [comment]
> What if it only processes if the meta url matches the fetched URL? So if it is different it has to fetch that one? I guess is that what you are saying?

I *think* we mean the same thing, yes. 🥴

In pseudo-code:

```
response = fetch(url)
if response.meta.url != url:
    response = fetch(response.meta.url)
use response.meta.url for hashing
```

In other words, make sure that `url =` and the actual URL we retrieved match.

**Comment by anths** [comment]
Do you allow for multiple redirects? If not, then you limit the application, say, preventing me from moving my feed address while also resolving the http/https ambiguity in the new one. Maybe that’s okay, and we just say “update your old forwards to be correct”, although there are cases where I might not have access to do so any more. Those are increasingly thin edge cases, but also totally plausible.

If you *do* allow multiple redirects, you provide a much more flexible system, but need to think about rewrite limits to deal with rewrite loops (among other things), intentional or not.

I’m inclined to say 1 forward only, fix your old forwards, but either way the spec needs to make that decision and state it clearly.

**Comment by JonLundy** [comment]
Even with a single redirect it should eventually get to the right place as long as the lookup tool is updating its internal feed location state each time.

**Comment by anths** [comment]
That doesn’t address loops. And I didn’t  realize updating the URL being followed was part of the proposal. That introduces a whole new set of questions. Like: are these more akin to temporary redirects or permanent redirect in HTTP, or more like a CNAME in DNS? What if the url line references a protocol I can’t speak? That’s not a problem if we’re just using it for the hash, But if the expectation is that my client will update its own internal records of who it follows, it’s likely to break more things.

I think vain’s pseudocode is pretty close as is: it doesn’t (and shouldn’t) update ‘url’, so we’ll continue to use the same one next time, it’s a single round of redirect, and treats the url line as a sort of combination temporary redirect and cname. Clients can make a note that they’re being redirected if they want, to let users update their records if they want.

It doesn’t address the case where there’s a redirect to a protocol I can’t access *and* the new target has a different url, but we’re in a pretty narrow edge case now.

**Comment by vain** [comment]
Alright, let’s refine the snippet/“proposal” above based on your feedback.

# Version 1 😒

```
hash_url = config.url

response = fetch(config.url)
if response.meta.url != config.url:
    alert user about mismatch
    response = fetch(response.meta.url)
    if response could be retrieved:
        hash_url = response.meta.url
    else:
        alert user about failure in retrieval

use hash_url for hashing
use response for displaying twts
```

Notes:

1.  Follow only one redirect. The user will be notified about this
    mismatch and they should update their config accordingly.
2.  Do not update the config automatically. This is fragile.
3.  Fall back to using `config.url` if we cannot retrieve the feed
    specified in `url =`.
4.  Display twts from the feed *after* following the redirect (as
    opposed to just using the original data).

Thoughts:

1.  My RSS feed reader automatically updates URLs in its database on
    redirects and that’s a terrible idea. Every now and then, I have to
    check my feed database for garbage: Sometimes, web servers send a
    *permanent redirect* during maintenance (which is wrong, but they
    still do it), so my feed reader now tries to access stuff like
    `https://example.com/maintenance.html` all the time.

    We really shouldn’t copy this model and let users decide instead.
    (The UI can still look like this: “Hey, it appears the feed URL has
    changed, do you want to update your config to $url? [Y/n]” *This*
    function could be allowed to follow multiple redirects, I guess.)
2.  Similarly, if we can’t access the URL from metadata, just alert the
    user.


# Version 2 😊

With all that said, I have to admit: I’m not really happy that I opened
a can of worms here. I’d rather not introduce an entirely new redirect
mechanism. Honestly, I’d be perfectly happy with this alternative
proposal and might even prefer it:

```
hash_url = config.url

response = fetch(config.url)
if response.meta.url != config.url:
    alert user about mismatch

use hash_url for hashing               (this is the original config.url)
use response for displaying twts
```

In other words, just *check* if `url =` matches the URL from the config
(as opposed to *make sure* it matches) and alert the user if it doesn’t.
If it doesn’t match, it’s an indicator that the user should update their
config, because they’re using an outdated/wrong URL to retrieve the
feed. The user should fix that, because they might break threading
otherwise.

**Comment by der-lyse** [comment]
I'm also in favor of version 2, it's much simpler and less prone to mess things up. If it turns out to be not enough, we can always upgrade it. But I reckon we should start with the easiest and least "invasive" thing.

However, if there are several `url` settings, what exactly should happen? If `config.url` matches any `response.meta.url`, does a warning occur or not? Is the first `response.meta.url` considered a special (like preferred) one? If so, does `config.url != reponse.meta.url[0]` trigger the warning?

**Comment by vain** [comment]
> However, if there are several url settings, what exactly should happen?

Mmmhhh … We’re kind of back to square one then. 🤔

I don’t know, the whole idea of serving the same feed over multiple protocols conflicts with twt hashes including the URL in the hashed data. Twt hashes basically dictate that a feed must only be served over one protocol/URL – or that we must blindly trust some metadata included within the feed.

My “version 2” tries to nudge all people to use the same URL. But then you might ask: “Why provide it via several protocols in the first place?”

So, yeah, “version 3” would be xuu’s original proposal (just use the first `url =` from the feed if it exists, use `config.url` otherwise).

**Comment by anths** [comment]
So “version 3” uses the first url in the file, if it exists, but doesn’t do redirects at all? I think that is fine, and simplifies the rest of these questions.

**Comment by vain** [comment]
Yeah. To clarify: “Version 3” would literally only use the URL to compute the twt hash – it would *not* try to actually retrieve the URL.

**Comment by der-lyse** [comment]
I'm all for version 3.

**Comment by JonLundy** [comment]
how does this sound..

1. Fetch feed from follow config.
2. If No URL found in the file use config url for backwards compat.
3. If url (or multiple) are included it must match the config url.
    3a. This should be validated on follow.
    3b. if its no longer included there should be a notification to update(re-follow) 
4. If multiple the FIRST is used for hashing.


So if i host my feed on HTTPS and Gemini as a secondary

FETCH HTTPS and see on list that https is on list. use that for hash.
FETCH Gemini and see that gemini is second on list. use HTTPS for hash.
FETCH HTTP get https redirected see its not on list. show twt that says "this feed is miss configured click here to refollow"

**Comment by prologic** [comment]
For what it's worth I think the case of providing a feed over multiple protocols is probably fairly low￼. Let's face it only technically minded geeky type folk would probably do this anyway 😂￼

**Comment by JonLundy** [comment]
it could be providing the feed over a backup HTTPS in case of maint

**Comment by anths** [comment]
Providing a feed over both http and https is probably at least common enough to take into account.

I dislike this option #4 because it adds additional requirements and doesn’t halo get to a canonical make, which is the point of this.

**Comment by JonLundy** [comment]
Providing the feed over http and https is no different from hosting via gopher and https. or UUCP and RFC1149..  They are both different protocols and should be treated as such. This option actually lowers the strictness of option 1 by allowing for whitelisted alternative host locations while keeping the first as the canonical feed address for hashing. 

If a user was to move from one location to another they would keep the original canonical address and the new hosted location would be listed second.. that way the hashes from old twts are not invalidated. 

It could be optioned that new comment lines that specify a feed url should change the feed used to hash like:
```
# url = proto://feed1.url
[TWT1]
[TWT2]
#url = proto://feed2.url
[TWT3]
```

where feed2 is canonical but TWT1 and TWT2 are hashed with feed1. This allows for potentially spoofing other twters unfortunately.  But as this system is already compromised due to a lack of crypto its probably all out of scope anyway.
